### ISSUE TITLE
Do not report as_conversions for coercions

### ISSUE THREAD

**Issue by SOF3** [rust-lang/rust-clippy#6428]
The use of `as` is discouraged when it is semantically unclear, such as conversion between numeric types. However, when it is used as a coercion, there is no better concise way to indicate the desired semantics.

For example, the following examples are legitimate:

```rust
fn identical_position<T>(a: &T, b: &T) -> bool {
    a as *const T == b as *const T
}

fn read(e: Enum) -> &mut dyn Read {
    match e {
        Enum::File(file) => file as &mut dyn Read,
        Enum::Cursor(cursor) => cursor as &mut dyn Read,
    }
}
```

I understand that `as_conversions` is set to `allow` for these reasons. However, i wish to set `warn(as_conversions)` at the crate-level to prevent future contributors from adding numeric casts to my project. It would be helpful to have a variant of `as_conversions` only for non-coercing casts, in particular, numeric casts.

**Comment by xfix** [comment]
The first example could be changed to use `std::ptr::eq` instead. Meanwhile, in the second example `as &mut dyn Read` shouldn't be necessary (as in, removal of `as &mut dyn Read` shouldn't make this code not compile).

**Comment by SOF3** [comment]
For the second example, there are some more complex cases where the compiler cannot infer it to be `&mut dyn` immediately. I cannot construct such an example by hand, but I think most people have run into this at least once.

**Comment by rail-rain** [comment]
@SOF3 Could you confirm that you need a _single_ lint that warns every uses of `as` except coercions, not a bunch of lints that triggers on different situations; already existing as the `cast_*` family for numeric casts and bonus `fn_to_numeric_cast*` + `char_lit_as_u8`. If I understand correctly, `as_conversions` is for people who want a simple lint to detect every use of `as`; independent from other `as` related lints.

**Comment by SOF3** [comment]
@rail-rain you mean separate lints such as `clippy::numeric_as_conversions`, `clippy::ptr_as_conversions`, etc.?

**Comment by rail-rain** [comment]
@SOF3 Um…I'm afraid not; the bunch of lints for specific cases (e.g. numeric casts that may truncate large values) already exist. You can find these by searching on "cast" [here](https://rust-lang.github.io/rust-clippy/master/index.html) (I couldn't find a way to set that as a query string). I wanted to check these lints do not fit your use case.

**Comment by SOF3** [comment]
oh yeah, I forgot those exist. should probably have used them instead.

**Comment by rail-rain** [comment]
In that case, do you still need this issue? Otherwise, could you close the issue please?

**Comment by rail-rain** [comment]
Thank you!
